### PR TITLE
Add xReadfile wrapper for reading small to medium size files

### PR DESCRIPTION
**Pull request by cgzones**
Inspired by proposed Linux syscall

Avoid file descriptor leaks like 4af8c63f

### CODE REVIEW

**Comment by BenBE** [comment]
Is it intentional that this does not allow to detect the file being too large?

What about `openat` support?

**Comment by cgzones** [comment]
> Is it intentional that this does not allow to detect the file being too large?

Yes.
One can check if the return value equals the passed `count`, then the file is probably bigger (except it has unlikely the size `count`).
But the current code this wrapper replaces does not check for this case and for procfs files we assess what maximum size they have under normal circumstances.

> What about openat support?

Will come if #321 gets accepted.

File: Compat.c [hunk 1]
```diff
@@ -117,3 +117,68 @@ int Compat_readlinkat(int dirfd,
...
+         buffer = ((char*)buffer) + res;
+         count -= (size_t)res;
+         alreadyRead += res;
+      }
+
+      if (count == 0 || res == 0) {
+         close(fd);
+         return alreadyRead;
+      }
+   }
```
Comment (BenBE) [inline Compat.c:147]: Extract to `xReadFD` for reuse in `xReadfileat` …

File: linux/LinuxProcessList.c [hunk 2]
```diff
@@ -108,28 +85,14 @@ static int sortTtyDrivers(const void* va, const void* vb) {
...
-   }
-   if (bufLen == 0) {
-      free(buf);
+   char buf[16384];
+   ssize_t r = xReadfile(PROCTTYDRIVERSFILE, buf, sizeof(buf) - 1);
+   if (r < 0)
       return;
-   }
+
+   buf[r] = '\0';
```
Comment (BenBE) [inline linux/LinuxProcessList.c:94]: What about making the NUL termination part of the xReadfile API? I.e. internally ensured before terminating that the byte following the last index written in buffer is forced to `\0`?

File: XUtils.c [hunk 3]
```diff
@@ -210,3 +211,46 @@ char* xStrndup(const char* str, size_t len) {
...
+
+      if (res > 0) {
+         buffer = ((char*)buffer) + res;
+         count -= (size_t)res;
+         alreadyRead += res;
+      }
+
+      if (count == 0 || res == 0) {
+         close(fd);
+         ((char*)buffer)[alreadyRead] = '\0';
```
Comment (BenBE) [inline XUtils.c:236]: Avoid buffer overflow:
```suggestion
         if (count)
            ((char*)buffer)[alreadyRead] = '\0';
```

**Review by BenBE — CHANGES_REQUESTED** [review]

File: XUtils.c (hunk 3 above)
Comment (cgzones) [reply XUtils.c:236]: there is previously the line `count--; // reserve one for null-terminator`.

Or should we check
```c
ssize_t xReadfile(const char* pathname, void* buffer, size_t count) {
   if (count == 0)
      return -EINVAL;
   ...
```
?

Comment (BenBE) [reply XUtils.c:236]: Unconditionally decrementing causes `MAX_INT` to be read, which is likely undesired as well. But decrementing the `count` on entry is quite a good solution. So return `-EINVAL` on `!count` before decrementing it. Didn't notice the decrement when I initially read through the source.

**Comment by cgzones** [comment]
Got a crash:

```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==34090==ERROR: AddressSanitizer: SEGV on unknown address 0x03e80000852a (pc 0x7ff3dd3bcfe1 bp 0x7ffe38245970 sp 0x7ffe38244f40 T0)
==34090==The signal is caused by a READ memory access.
    #0 0x7ff3dd3bcfe1 in raise nptl/../sysdeps/unix/sysv/linux/raise.c:51:1
    #1 0x4ed427 in CRT_handleSIGSEGV /home/christian/Coding/workspaces/htop/CRT.c:894:4
    #2 0x7ff3dd3bd13f  (/lib/x86_64-linux-gnu/libpthread.so.0+0x1413f)
    #3 0x7ff3dd254082  string/../sysdeps/x86_64/multiarch/../strchr.S:31
    #4 0x43cab8 in strchr (/home/christian/Coding/workspaces/htop/htop+0x43cab8)
    #5 0x57632b in LinuxProcessList_readStatFile /home/christian/Coding/workspaces/htop/linux/LinuxProcessList.c:351:18
    #6 0x56f201 in LinuxProcessList_recurseProcTree /home/christian/Coding/workspaces/htop/linux/LinuxProcessList.c:1330:13
    #7 0x567c6a in ProcessList_goThroughEntries /home/christian/Coding/workspaces/htop/linux/LinuxProcessList.c:1905:4
    #8 0x535c05 in ProcessList_scan /home/christian/Coding/workspaces/htop/ProcessList.c:346:4
    #9 0x53e803 in checkRecalculation /home/christian/Coding/workspaces/htop/ScreenManager.c:108:7
    #10 0x53bfc8 in ScreenManager_run /home/christian/Coding/workspaces/htop/ScreenManager.c:159:10
    #11 0x4fefe7 in main /home/christian/Coding/workspaces/htop/htop.c:336:4
    #12 0x7ff3dd1ddcc9 in __libc_start_main csu/../csu/libc-start.c:308:16
    #13 0x4278f9 in _start (/home/christian/Coding/workspaces/htop/htop+0x4278f9)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV nptl/../sysdeps/unix/sysv/linux/raise.c:51:1 in raise
==34090==ABORTING
```

Checking on master...

**Review by BenBE — APPROVED** [review]

**Pull request merged**